Create scaling objects instead of hpa

Richard87 opened the pull request.

nilsgstrabo commented on pkg/apis/deployment/scaledobject.go:
```diff
@@ -0,0 +1,178 @@
+	return deploy.kubeutil.ApplyScaledObject(ctx, namespace, scaler)
+}
+
+func (deploy *Deployment) garbageCollectDeprecatedHPAs(ctx context.Context) error {
```
This function will delete all HPAs, also HPAs controlled by ScaledObjects. We should only delete HPAs not controlled by ScaledObjects. We can do that by filtering on labels (set by ScaledOBject).
Example of HPA labels:
labels:
    app.kubernetes.io/managed-by: keda-operator
    app.kubernetes.io/name: keda-hpa-server
    app.kubernetes.io/part-of: server
    app.kubernetes.io/version: 2.14.0
    radix-app: radix-api
    radix-component: server
    scaledobject.keda.sh/name: server

Richard87 replied:
Fixed: Ignorerer HPA'en hvis `owner` ikke er `RadixDeployment`, da får vi slettet alle gamle hpas, men ikke de nye:

```golang
// If owner reference is *not* RadixDeployment, skip deleting it
if len(hpa.OwnerReferences) == 0 {
	continue
}
if owner := hpa.OwnerReferences[0]; owner.Kind != radixv1.KindRadixDeployment {
	continue
}
```

nilsgstrabo reviewed (changes requested):
Added some suggestions and comments.
We must also add ScaledObject to the radix-operator clusterrole (all verbs), and to the radix-api and radix-app-admin-envs clusterroles (get, list, watch) if radix-api needs to read the ScaledObject.

Richard87 commented on pkg/apis/utils/scaler.go:
```diff
@@ -14,3 +15,13 @@ func GetHpaMetric(hpa *v2.HorizontalPodAutoscaler, resourceName v1.ResourceName)
 }
+
+// GetScaleTrigger returns the trigger spec for the given name  and type or nil if not found
+func GetScaleTrigger(hpa *v1alpha1.ScaledObject, triggerName, triggerType string) *v1alpha1.ScaleTriggers {
```
Fixed

Richard87 commented on pkg/apis/radixvalidators/validate_ra_test.go:
```diff
@@ -1749,7 +1778,7 @@ func Test_HPA_Validation(t *testing.T) {
 			},
 			false,
-			false,
+			[]error{radixvalidators.ErrMinReplicasGreaterThanMaxReplicas, radixvalidators.ErrMinReplicasGreaterThanMaxReplicas},
```
Fixed, was a bug :)

Richard87 commented on pkg/apis/radixvalidators/validate_ra_test.go:
```diff
@@ -1589,24 +1589,24 @@ func Test_ValidationOfVolumeMounts_Errors(t *testing.T) {
 			func(ra *radixv1.RadixApplication) {},
 			true,
-			true,
+			[]error{},
```
Nice! Fixed :)

Richard87 commented on pkg/apis/radixvalidators/validate_ra_test.go:
```diff
@@ -1589,24 +1589,24 @@ func Test_ValidationOfVolumeMounts_Errors(t *testing.T) {
-		isErrorNil bool
+		name     string
+		updateRA updateRAFunc
+		isValid  bool
+		isErrors []error
 	}{
```
Agreed, with `isErrors = nil`, its pretty clear... fixed!

Richard87 commented on pkg/apis/deployment/kubedeployment.go:
```diff
@@ -21,15 +21,15 @@ import (
 )
 
-func (deploy *Deployment) createOrUpdateDeployment(ctx context.Context, deployComponent v1.RadixCommonDeployComponent) error {
+func (deploy *Deployment) recconcileDeployment(ctx context.Context, deployComponent v1.RadixCommonDeployComponent) error {
```
fixed

Richard87 commented:
> Added some suggestions and comments. We must also add ScaledObject to the radix-operator clusterrole (all verbs), and to the radix-api and radix-app-admin-envs clusterroles (get, list, watch) if radix-api needs to read the ScaledObject.

Fixed, added `get,list,watch` to app admins and app readers (in `-envs` cluster roles), and `get,list,create,update,patch,delete,watch` to  Radix Operator

Richard87 commented on pkg/apis/radixvalidators/validate_ra_test.go:
```diff
@@ -1731,15 +1759,16 @@ func Test_HPA_Validation(t *testing.T) {
 			false,
-			false,
+			// TODO: Maybe this validation should not return ErrMinReplicasGreaterThanMaxReplicas (it happens because minReplicas defaults to 1
+			[]error{radixvalidators.ErrMinReplicasGreaterThanMaxReplicas, radixvalidators.ErrMinReplicasGreaterThanMaxReplicas},
```
Fixed, was a bug :)

Richard87 commented on pkg/apis/deployment/scaledobject.go:
```diff
@@ -0,0 +1,178 @@
+			MinReplicaCount: minReplicas,
+			MaxReplicaCount: pointers.Ptr(maxReplicas),
+			PollingInterval: pointers.Ptr[int32](30), // Default
+			Advanced:        &kedav1.AdvancedConfig{RestoreToOriginalReplicaCount: true},
```
Will merge and release this before the Operator https://github.com/equinor/radix-api/pull/625
